backend/packages/Upgrade/src/api/services/MoocletRewardsService.ts

```diff
@@ -0,0 +1,252 @@
+  constructor(
+    @InjectRepository()
+    private moocletExperimentRefRepository: MoocletExperimentRefRepository,
+    @InjectRepository()
+    private individualEnrollmentRepository: IndividualEnrollmentRepository,
+    private metricService: MetricService,
+    private moocletDataService: MoocletDataService
+  ) {}
+
+  public getRewardMetricQuery(rewardMetricKey: string): QueryValidator {
+    return {
+      id: uuid(),
+      name: 'Success Rate',
+      query: {
+        operationType: 'percentage',
+        compareFn: '=',
+        compareValue: BinaryRewardMetricAllowedValue.SUCCESS,
+      },
+      metric: {
+        key: rewardMetricKey,
+      },
+      repeatedMeasure: REPEATED_MEASURE.mostRecent,
+    };
+  }
+
+  public async createAndSaveRewardMetric(
+    rewardMetricKey: string,
+    context: string,
+    logger: UpgradeLogger
+  ): Promise<Metric[]> {
+    const metric = {
+      id: uuid(),
+      metric: rewardMetricKey,
+      datatype: IMetricMetaData.CATEGORICAL,
+      allowedValues: [BinaryRewardMetricAllowedValue.SUCCESS, BinaryRewardMetricAllowedValue.FAILURE],
+    };
+
+    return this.metricService.saveAllMetrics([metric], [context], logger);
+  }
+
+  public async parseLogsAndSendPotentialRewards(
+    user: RequestedExperimentUser,
+    logs: ILogInput[],
+    logger: UpgradeLogger
+  ): Promise<void> {
+    // First see if any simple metrics have valid reward values
+    const validSimpleMetricAttributes = this.gatherValidRewardMetrics(logs);
+
+    if (!validSimpleMetricAttributes.length) {
+      // no need to log anything here, just quit early as this is the most common case
+      return;
+    }
+
+    try {
+      // If we have valid metric rewards, only then check for active mooclet experiment refs
+      const moocletExperimentRefs = await this.getAllActiveMoocletExperimentRefs(logger);
+
+      if (!moocletExperimentRefs.length) {
+        logger.warn({
+          message: 'Reward metrics were logged, but no active mooclet experiment refs found.',
+          user,
+        });
+        return;
+      }
+
+      // Find each experiment match by reward metric key
+      const experimentsMatchingLoggedRewardKey = this.findExperimentByRewardMetricKey(
+        moocletExperimentRefs,
+        validSimpleMetricAttributes
+      );
+
+      if (!experimentsMatchingLoggedRewardKey.length) {
+        logger.warn({
+          message: 'Reward metrics were logged, but no active experiments matched the rewardMetricKeys',
+          moocletExperimentRefs,
+          validSimpleMetricAttributes,
+        });
+        return;
+      }
+
+      // For each match, find the user's condition, map to version id, and create reward object
+      experimentsMatchingLoggedRewardKey.forEach(async (experimentRewardPair) => {
+        const { moocletExperimentRef, rewardMetricValue } = experimentRewardPair;
+        const enrollment = await this.findEnrolledCondition(user.id, moocletExperimentRef.experimentId, logger);
```

Comment:
A db call is happening for each experiment. We can fetch all the enrollment documents in a single call.